cargo-vendor: 0.1.13 -> 0.1.23

akru opened the pull request:
###### Motivation for this change

Upgrade cargo-vendor to 0.1.23 is required for support build edition 2018 crates.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

akru commented:
@GrahamcOfBorg build cargo-vendor

andir commented on pkgs/build-support/rust/default-crate-overrides.nix:
```diff
@@ -14,7 +14,7 @@ in
       ++ stdenv.lib.optionals stdenv.isDarwin [ CoreFoundation libiconv ];
     # TODO: buildRustCrate seems to use incorrect default inference
-    crateBin = [ {  name = "cargo"; path = "src/bin/cargo.rs"; } ];
+    crateBin = [ {  name = "cargo"; path = "src/bin/cargo/main.rs"; } ];
```
You can remove the `path` attribute. Since fc5e5950031d8af57f8b5b9e55187f3e4cb4f063 we are doing the right thing(tm) within `buildRustCrate`.

akru replied:
You are right, fixed

akru commented:
Unfortunately with big crate deps list (like a `parity`) `cargo-vendor` has segfault in `libcurl.so`. WIP

andir commented:
I am only seeing

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/glsy9j0x5j23sq1rqnazs1jvd9hd7mxl-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "installPhase" }
installing
    Updating git repository `https://github.com/cheme/heapsize.git`
    Updating crates.io index
    Updating git repository `https://github.com/paritytech/rust-secp256k1`
    Updating git repository `https://github.com/paritytech/app-dirs-rs`
    Updating git repository `https://github.com/paritytech/bn`
    Updating git repository `https://github.com/paritytech/jsonrpc.git`
    Updating git repository `https://github.com/paritytech/hidapi-rs`
    Updating git repository `https://github.com/paritytech/libusb-rs`
    Updating git repository `https://github.com/paritytech/trezor-sys`
    Updating git repository `https://github.com/paritytech/rust-ctrlc.git`
    Updating git repository `https://github.com/paritytech/libusb-sys`
    Updating git repository `https://github.com/tomusdrw/ws-rs`
 Downloading crates ...
  Downloaded digest v0.7.6
  Downloaded crossbeam-deque v0.5.2
Traceback (most recent call last):
  File "/nix/store/7qnming9jgiqb6hk2pqfmgp4dn60vs5r-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 42, in <module>
    main()
  File "/nix/store/7qnming9jgiqb6hk2pqfmgp4dn60vs5r-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 17, in main
    assert list(data.keys()) == ["source"]
AssertionError
```

When running `nix-build -I nixpkgs=. -A parity.cargoDeps --check` on this PR.
Is that what you are seeing?

akru commented:
```
Traceback (most recent call last):
  File "/nix/store/7qnming9jgiqb6hk2pqfmgp4dn60vs5r-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 42, in <module>
    main()
  File "/nix/store/7qnming9jgiqb6hk2pqfmgp4dn60vs5r-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 17, in main
    assert list(data.keys()) == ["source"]
AssertionError
```
@andir This assertion is you catch because `cargo vendor` program fails with segmentation fault. You can see it in `dmesg` log.

dtzWill commented:
FWIW this seems to be working for me? If I'm the only one NOT seeing segfaults I can investigate further... I am building on top of staging's rust bump to 1.33-- no clue if that's relevant :).

andir commented:
@dtzWill how did you test it? I had to run `--check` on the `cargoDeps` attribute to actually re-run the whole cargo-vendor thing.

dtzWill commented:
On Fri, 08 Mar 2019 12:06:09 -0800, Andreas Rammhold <notifications@github.com> wrote:
> @dtzWill how did you test it? I had to run `--check` on the `cargoDeps` attribute to actually re-run the whole cargo-vendor thing.

Yep, with `--check` using the command for `parity` you posted.

I also used it in packaging up a newer (git) version of `fractal` \o/
which was what led me to discovering this PR :).

Let's see what behavior this has for other folks, maybe?

> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/57017#issuecomment-471059453Non-text part: text/html

andir commented:
@GrahamcOfBorg build parity.cargoDeps 

Does ofBorg also have --check?

andir commented:
@dtzWill I tried to again on a newly installed NixOS packet.net machine. Cloned nixpkgs and ran `nix-build -I nixpkgs=. '<nixpkgs>' -A parity.cargoDeps --check`. Segfaults like on all my other machines.

What differences are there between our systems?

Laaas commented:
FWIW I'm not seeing any failures with the command andir posted.

flokli commented:
I was able to reproduce this after a second try:

```
⇒  nix-build -I nixpkgs=. -A parity.cargoDeps --check
checking outputs of '/nix/store/fvqwb3q1yf8kv2h4ld8v0ij92v51g7lw-parity-2.2.9-vendor.drv'...
unpacking sources
unpacking source archive /nix/store/glsy9j0x5j23sq1rqnazs1jvd9hd7mxl-source
source root is source
patching sources
installing
    Updating git repository `https://github.com/cheme/heapsize.git`
    Updating crates.io index
    Updating git repository `https://github.com/paritytech/rust-secp256k1`
    Updating git repository `https://github.com/paritytech/app-dirs-rs`
    Updating git repository `https://github.com/paritytech/bn`
    Updating git repository `https://github.com/paritytech/jsonrpc.git`
    Updating git repository `https://github.com/paritytech/hidapi-rs`
    Updating git repository `https://github.com/paritytech/libusb-rs`
    Updating git repository `https://github.com/paritytech/trezor-sys`
    Updating git repository `https://github.com/paritytech/rust-ctrlc.git`
    Updating git repository `https://github.com/paritytech/libusb-sys`
    Updating git repository `https://github.com/tomusdrw/ws-rs`
 Downloading crates ...
  Downloaded url v1.7.1
  Downloaded indexmap v1.0.2
Traceback (most recent call last):
  File "/nix/store/7qnming9jgiqb6hk2pqfmgp4dn60vs5r-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 42, in <module>
    main()
  File "/nix/store/7qnming9jgiqb6hk2pqfmgp4dn60vs5r-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 17, in main
    assert list(data.keys()) == ["source"]
AssertionError
builder for '/nix/store/fvqwb3q1yf8kv2h4ld8v0ij92v51g7lw-parity-2.2.9-vendor.drv' failed with exit code 1
error: build of '/nix/store/fvqwb3q1yf8kv2h4ld8v0ij92v51g7lw-parity-2.2.9-vendor.drv' failed
```
dmesg:
```
[333635.128291] cargo-vendor[4457]: segfault at 0 ip 00007ffff77c430e sp 00007ffffffed820 error 4 in libc-2.27.so[7ffff7778000+13d000]
[333635.128296] Code: 2e 0f 1f 84 00 00 00 00 00 0f 1f 00 41 57 41 56 41 55 41 54 55 53 48 89 f3 48 0f af da 48 83 ec 08 48 85 db 0f 84 22 01 00 00 <8b> 01 49 89 fc 49 89 f2 49 89 c9 25 00 80 00 00 75 59 4c 8b 81 88
```

dtzWill commented:
To be clear, I've been using this on top of my normal branch and not the exact tree here, and there are many reasons I'd be seeing different behavior-- but MOSTLY the bits that turn out well I send upstream so I don't think I'm sitting on anything relevant.

The most obvious difference is I'm using the rust 1.33 updates from our staging branch-- and I don't know what changed in 1.33 but "fixing things that cause crashes" sounds plausible enough to check it out ;).

When my builders have a chance I'll try the tree here and maybe on a few machines... heisenbug is lame!

akru committed:
cargo-vendor: 0.1.13 -> 0.1.23

akru commented:
As I see it fixed in latest `cargo-vendor`.

akru commented:
@GrahamcOfBorg build cargo-vendor

sondr3 commented:
This would be really nice to have, just ran into issues with the old version of `cargo-vendor` with `lto = "thin"`, fingers crossed this works out :smile:

lilyball commented:
What's the current status of this? It looks like I can't update `ffsend` right now due to the current `cargo-vendor` not understanding a dependency of the form `clip = { version = "0.5", optional = true, package = "clipboard" }` (e.g. where the actual package name doesn't match the dependency name). I'm hoping this update will fix that.

Incidentally, I can't build this on macOS. I tried merging it into current `master` and building, and `rust_cargo-0.35.0` fails to build, saying it can't find the Security framework (which according to the .drv is not listed as a buildInput).

lilyball commented on pkgs/build-support/rust/default-crate-overrides.nix:
```diff
@@ -13,15 +13,15 @@ in
   cargo = attrs: {
     buildInputs = [ openssl zlib curl ]
       ++ stdenv.lib.optionals stdenv.isDarwin [ CoreFoundation libiconv ];
```
It looks like my macOS build error can be fixed with

```suggestion
      ++ stdenv.lib.optionals stdenv.isDarwin [ CoreFoundation Security libiconv ];
```

Of course, then I get the same error for cargo-vendor itself, so the same fix needs to be applied there too.

lilyball commented on pkgs/build-support/rust/default-crate-overrides.nix:
```diff
@@ -13,15 +13,15 @@ in
   };
 
   cargo-vendor = attrs: {
     buildInputs = [ openssl zlib curl ];
```
```suggestion
    buildInputs = [ openssl zlib curl ]
      ++ stdenv.lib.optionals stdenv.isDarwin [ Security ];
```

andir commented:
I've been trying to reproduce the segfault we did observe with this rebased onto  currentmaster. After 6 rebuilds there hasn't been a single segfault. Currently trying to reproduce it on another machine but that also succeeded for a few times. I hope this isn't due to the host being on nixos-19.03 now...

@lilyball can you apply the mentioned fix to cargo to figure out if this package (and the users) work on master for Darwin? I do not own any of that hardware :/

lilyball commented:
@andir I already applied the two fixes I suggested and successfully built `cargo-vendor` (and `ffsend-0.2.45`) on Darwin.

andir committed:
cargo: fix build on Darwin

This is supposedly fixing the build of the cargo crate on Drawin [1].

[1] https://github.com/NixOS/nixpkgs/pull/57017#pullrequestreview-228868016

andir committed:
cargo-vendor: fix build on Darwin

This is supposedly fixing the build of the cargo crate on Drawin [1].

[1] https://github.com/NixOS/nixpkgs/pull/57017#pullrequestreview-228868016

andir commented:
@lilyball does this look alright now?

lilyball commented:
@andir Yes, it builds just fine on Darwin now.
